### ISSUE TITLE
Investigate and fix/resolve memcpy compiler warnings in OMRMachine.cpp

### ISSUE THREAD

**Issue by mstoodle** [eclipse/omr#840]
Recently caught in a fast scrolling build log on OSX:

...
clang++ -std=c++0x -fno-rtti -fno-threadsafe-statics -Wno-deprecated -Wno-enum-compare -Wno-invalid-offsetof -Wno-write-strings -O3 -pthread -fomit-frame-pointer -fasynchronous-unwind-tables -Wreturn-type -fno-dollars-in-identifiers -Wno-parentheses -Werror=header-guard   -DBITVECTOR_BIT_NUMBERING_MSB -DUT_DIRECT_TRACE_REGISTRATION -DJITTEST -DJITBUILDER_SPECIFIC -DPROD_WITH_ASSUMES -DTR_HOST_X86 -DTR_HOST_64BIT -DBITVECTOR_64BIT -DOSX -DTR_TARGET_X86 -DTR_TARGET_64BIT -DSUPPORTS_THREAD_LOCAL -D_LONG_LONG -I'../include_core' -I'../jitbuilder/x/amd64' -I'../jitbuilder/x' -I'../jitbuilder' -I'../compiler/x/amd64' -I'../compiler/x' -I'../compiler' -I'..' -MMD -MF ../build/compiler/x/codegen/OMRMachine.o.depend.mk -MT ../build/compiler/x/codegen/OMRMachine.o -MP -o ../build/compiler/x/codegen/OMRMachine.o -c ../compiler/x/codegen/OMRMachine.cpp
../compiler/x/codegen/OMRMachine.cpp:1778:14: warning: destination for this 'memcpy' call is a pointer to dynamic class 'RealRegister';
      vtable pointer will be overwritten [-Wdynamic-class-memaccess]
      memcpy(registerFileClone[i], registerFile[i], sizeof(TR::RealRegister));
      ~~~~~~ ^
../compiler/x/codegen/OMRMachine.cpp:1778:14: note: explicitly cast the pointer to silence this warning
      memcpy(registerFileClone[i], registerFile[i], sizeof(TR::RealRegister));
             ^
             (void*)
../compiler/x/codegen/OMRMachine.cpp:1926:14: warning: destination for this 'memcpy' call is a pointer to dynamic class 'RealRegister';
      vtable pointer will be overwritten [-Wdynamic-class-memaccess]
      memcpy(registerFileClone[i], _registerFile[i], sizeof(TR::RealRegister));
      ~~~~~~ ^
../compiler/x/codegen/OMRMachine.cpp:1926:14: note: explicitly cast the pointer to silence this warning
      memcpy(registerFileClone[i], _registerFile[i], sizeof(TR::RealRegister));
             ^
             (void*)
../compiler/x/codegen/OMRMachine.cpp:1931:11: warning: destination for this 'memcpy' call is a pointer to dynamic class 'RealRegister';
      vtable pointer will be overwritten [-Wdynamic-class-memaccess]
   memcpy(registerFileClone[TR::RealRegister::vfp], _registerFile[TR::RealRegister::vfp], sizeof(TR::RealRegister));
   ~~~~~~ ^
../compiler/x/codegen/OMRMachine.cpp:1931:11: note: explicitly cast the pointer to silence this warning
   memcpy(registerFileClone[TR::RealRegister::vfp], _registerFile[TR::RealRegister::vfp], sizeof(TR::RealRegister));
          ^
          (void*)
../compiler/x/codegen/OMRMachine.cpp:1969:14: warning: destination for this 'memcpy' call is a pointer to dynamic class 'RealRegister';
      vtable pointer will be overwritten [-Wdynamic-class-memaccess]
      memcpy(_registerFile[i], registerFileCopy[i], sizeof(TR::RealRegister));
      ~~~~~~ ^
../compiler/x/codegen/OMRMachine.cpp:1969:14: note: explicitly cast the pointer to silence this warning
      memcpy(_registerFile[i], registerFileCopy[i], sizeof(TR::RealRegister));
             ^
             (void*)
../compiler/x/codegen/OMRMachine.cpp:1986:11: warning: destination for this 'memcpy' call is a pointer to dynamic class 'RealRegister';
      vtable pointer will be overwritten [-Wdynamic-class-memaccess]
   memcpy(_registerFile[TR::RealRegister::vfp], registerFileCopy[TR::RealRegister::vfp], sizeof(TR::RealRegister));
   ~~~~~~ ^
../compiler/x/codegen/OMRMachine.cpp:1986:11: note: explicitly cast the pointer to silence this warning
   memcpy(_registerFile[TR::RealRegister::vfp], registerFileCopy[TR::RealRegister::vfp], sizeof(TR::RealRegister));
          ^
          (void*)
../compiler/x/codegen/OMRMachine.cpp:2003:17: warning: destination for this 'memcpy' call is a pointer to dynamic class 'Register'; vtable
      pointer will be overwritten [-Wdynamic-class-memaccess]
         memcpy(registerAssociationsClone[i], _registerAssociations[i], sizeof(TR::Register));
         ~~~~~~ ^
../compiler/x/codegen/OMRMachine.cpp:2003:17: note: explicitly cast the pointer to silence this warning
         memcpy(registerAssociationsClone[i], _registerAssociations[i], sizeof(TR::Register));
                ^
                (void*)
../compiler/x/codegen/OMRMachine.cpp:2015:14: warning: destination for this 'memcpy' call is a pointer to dynamic class 'Register'; vtable
      pointer will be overwritten [-Wdynamic-class-memaccess]
      memcpy(registerAssociationsClone[TR::RealRegister::vfp], _registerAssociations[TR::RealRegister::vfp], sizeof(TR::Register));
      ~~~~~~ ^
../compiler/x/codegen/OMRMachine.cpp:2015:14: note: explicitly cast the pointer to silence this warning
      memcpy(registerAssociationsClone[TR::RealRegister::vfp], _registerAssociations[TR::RealRegister::vfp], sizeof(TR::Register));
             ^
             (void*)
7 warnings generated.


Seems like a warning worth checking into and either fixing or following the advice if that's somehow the right answer.

**Comment by mstoodle** [comment]
OMR::RealRegister appears to have virtual members (block, unblock, getRegister, getRealRegister).

Virtual members leads one to consider the possibility of subclasses of TR::RealRegister.

The existence of subclasses makes one very worried about code that uses memcpy to copy one TR::RealRegister object over another TR::RealRegister object (with, you guessed it, sizeof(TR::RealRegister) ).

<squinting dubiously> Maybe the subclasses don't add any instance fields?

**Comment by rwy0717** [comment]
The subclass RegisterPair definitely adds fields. I'm going to guess this works because the objects being memcpy'd are always TR::Register, not any of the subclasses.

**Comment by 0xdaryl** [comment]
Should be fixed by #3604.
